checker/src/org/checkerframework/checker/signedness/SignednessVisitor.java

```diff
@@ -23,6 +22,124 @@ public SignednessVisitor(BaseTypeChecker checker) {
+
+        return kind == Kind.AND || kind == Kind.OR;
+    }
+
+    /** @return true iff expr is a literal */
+    private boolean isLiteral(ExpressionTree expr) {
+        return expr instanceof LiteralTree;
+    }
+
+    /**
+     * @param obj either an Integer or a Long
+     * @return the long value of obj
+     */
+    private long getLong(Object obj) {
+        if (obj instanceof Integer) {
+            Integer intObj = (Integer) obj;
+            return intObj.longValue();
+        } else {
+            return (long) obj;
+        }
+    }
+
+    /**
+     * Given a masking operation of the form {@code x & maskLit} or {@code x | maskLit}, return true
+     * iff the masking operation results in the same output regardless of the value of the numBits
+     * most significant bits of x. This is if the numBits most significant bits of mask are 0 for
+     * AND, and 1 for OR. For example, assuming that numBits is 4, the following is true about AND
+     * and OR masks:
+     *
+     * <p>{@code x & 0x0... == 0x0...;} {@code x | 0xF... == 0xF...;}
+     *
+     * @param maskKind the kind of mask (AND or OR)
+     * @param numBitsLit the LiteralTree whose value is numBits
+     * @param maskLit the LiteralTree whose value is mask
+     * @return true iff the numBits most significant bits of mask are 0 for AND, and 1 for OR
+     */
+    private boolean maskIgnoresMSB(Kind maskKind, LiteralTree numBitsLit, LiteralTree maskLit) {
+        long numBits = getLong(numBitsLit.getValue());
+        long mask = getLong(maskLit.getValue());
+
+        // If maskLit was an int, then shift mask so that the numBits most significant bits are in the right position
+        if (maskLit.getKind() != Kind.LONG_LITERAL) {
+            mask <<= 32;
+        }
+
+        // Shift the numBits most significant bits to become the numBits least significant bits, zeroing out the rest
+        mask >>>= (64 - numBits);
+        if (maskKind == Kind.AND) {
+            // Check that the numBits most significant bits of the mask were 0
+            return mask == 0;
+        } else if (maskKind == Kind.OR) {
+            // Check that the numBits most significant bits of the mask were 1
+            return mask == (1 << numBits) - 1;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Determines if a right shift operation, {@code >>} or {@code >>>}, is masked with a masking
+     * operation of the form {@code shiftOp & maskLit} or {@code shiftOp | maskLit} such that the
+     * mask renders the shift signedness irrelevent by destroying the bits introduced by the shift.
+     * For example, the following pairs of right shifts on {@code byte b} both produce the same
+     * results under any input, because of their masks:
+     *
+     * <p>{@code (b >> 4) & 0x0F == (b >>> 4) & 0x0F;} {@code (b >> 4) | 0xF0 == (b >>> 4) | 0xF0;}
+     *
+     * @param shiftOp a right shift operation: {@code >>} or {@code >>>}
```

Comment:
This comment isn't right:  shiftOp is something like `expr1 >> expr2`, not just `>>`, right?

Separately, "shiftOp" sounds like an operator rather than an expression or tree; I would rename the variable.  The same goes for maskOp, which sounds to me like the operator itself.